src/Symfony/Component/Lock/Store/MongoDbStore.php

```diff
@@ -0,0 +1,329 @@
+     *
+     * CAUTION: The locked resouce name is indexed in the _id field of the
+     * lock collection.
+     * An indexed field's value in MongoDB can be a maximum of 1024 bytes in
+     * length inclusive of structural overhead.
+     *
+     * @see https://docs.mongodb.com/manual/reference/limits/#Index-Key-Limit
+     *
+     * CAUTION: This store relies on all client and server nodes to have
+     * synchronized clocks for lock expiry to occur at the correct time.
+     * To ensure locks don't expire prematurely; the lock TTL should be set
+     * with enough extra time to account for any clock drift between nodes.
+     * @see self::createTTLIndex()
+     *
+     * writeConcern, readConcern and readPreference are not specified by
+     * MongoDbStore meaning the collection's settings will take effect.
+     * @see https://docs.mongodb.com/manual/applications/replication/
+     */
+    public function __construct(Client $mongo, array $options, float $initialTtl = 300.0)
+    {
+        if (!isset($options['database'])) {
+            throw new InvalidArgumentException(
+                'You must provide the "database" option for MongoDBStore'
+            );
+        }
+
+        $this->mongo = $mongo;
+
+        $this->options = array_merge(array(
+            'collection' => 'lock',
+        ), $options);
+
+        $this->initialTtl = $initialTtl;
+    }
+
+    /**
+     * Create a TTL index to automatically remove expired locks.
+     *
+     * This should be called once during database setup.
+     *
+     * Alternatively the TTL index can be created manually:
+     *
+     *  db.lock.ensureIndex(
+     *      { "expires_at": 1 },
+     *      { "expireAfterSeconds": 0 }
+     *  )
+     *
+     * Please note, expires_at is based on the application server. If the
+     * database time differs; a lock could be cleaned up before it has expired.
+     * Set a positive expireAfterSeconds to account for any time drift between
+     * application and database server.
+     *
+     * A TTL index MUST BE used on MongoDB 2.2+ to automatically clean up expired locks.
+     *
+     * @see http://docs.mongodb.org/manual/tutorial/expire-data/
+     *
+     * @return string The name of the created index
+     *
+     * @throws UnsupportedException          if options are not supported by the selected server
+     * @throws MongoInvalidArgumentException for parameter/option parsing errors
+     * @throws DriverRuntimeException        for other driver errors (e.g. connection errors)
+     */
+    public function createTTLIndex(): string
+    {
+        $keys = array(
+            'expires_at' => 1,
+        );
+
+        $options = array(
+            'expireAfterSeconds' => 0,
+        );
+
+        return $this->getCollection()->createIndex($keys, $options);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws LockStorageException if failed to save to storage (fallback exception)
```

Comment:
From the documentation, it seems like `LockStorageException` should only be used when getting an error trying to manipulate an already existing resource in the store.

/cc - @jderusse in case he has better insight between the usage of `LockStorageException` or `LockAcquiringException`

Reply:
I'm just looking at `Lock::acquire` and `Lock::refresh`; both have a `catch (\Exception $e)` re-throwing as `LockAcquiringException`. I have therefore just removed my `} catch (Exception $e) {`'s in `MongoDbStore::save` and `MongoDbStore::putOffExpiration` to allow `Lock` to maintain exception consistency.